### PR TITLE
WT-10873 Pin the page we found during prepared update resolution.

### PR DESCRIPTION
**Pull request by luke-pearson**
Given that the transaction modification list is sorted it is likely to be relevant for the next modification. Additionally by pinning it we prevent it from being evicted and ensure that we can find the same update chain again.

This ~~is~~ was a draft PR, it does fix the issue but I am concerned about the change in cursor semantics.

### CODE REVIEW

File: src/btree/bt_cursor.c [hunk 1]
```diff
@@ -629,23 +629,19 @@ __wt_btcur_search_prepared(WT_CURSOR *cursor, WT_UPDATE **updp)
...
-    btree = CUR2BT(cbt);
-    upd = NULL; /* -Wuninitialized */
+    bool leaf_found;
 
     /*
-     * Not calling the cursor initialization functions, we don't want to be tapped for eviction nor
-     * do we want other standard cursor semantics like snapshots, just discard the hazard pointer
-     * from the last operation. This also depends on the fact we're not setting the cursor's active
-     * flag, this is really a special chunk of code and not to be modified without careful thought.
+     * We don't need to know if we found the key on the ref we provide but the search function
```
Comment (luke-pearson) [inline src/btree/bt_cursor.c:635]: The previous version of this comment said the following: "this is really a special chunk of code and not to be modified without careful thought.". I have removed all logic it refers to which is admittedly a little concerning but I cannot determine what exactly we achieved by resetting the cursor here. One thing a reset does is it calls `__cursor_pos_clear`, I could in theory call that directly that way we keep our hazard pointer.

Before a `__cursor_reset` was added here a `__wt_cursor_func_init` was here which has additional semantics on top of the reset. I am unsure what exactly we need.

Comment (luke-pearson) [reply src/btree/bt_cursor.c:635]: From the ticket it appears we don't want the cursor to get trapped in eviction by avoiding setting the cursor's active flag.

Comment (luke-pearson) [reply src/btree/bt_cursor.c:635]: I could F_CLR the active flag here.

Comment (agorrod) [reply src/btree/bt_cursor.c:635]: I don't think this needs anything further to get the semantics originally desired. The change in WT-5480 followed soon after WT-5387. Before WT-5387, the code would open an entirely new cursor for each op being resolved, and the goal of this was to re-use cursors. With 5387 cursors were reused, but the could be co-opted into eviction between usages. 
OTOH: I'm concerned that this code is leaking hazard pointers somehow - we've seen some signal in the field that hazard pointers can be acquired in surprising ways..

File: src/btree/bt_cursor.c [hunk 2]
```diff
@@ -629,23 +629,19 @@ __wt_btcur_search_prepared(WT_CURSOR *cursor, WT_UPDATE **updp)
...
-    WT_RET(__cursor_reset(cbt));
+    leaf_found = false;
+    *updp = upd = NULL; /* -Wuninitialized */
+    cbt = (WT_CURSOR_BTREE *)cursor;
+    btree = CUR2BT(cbt);
 
-    WT_RET(btree->type == BTREE_ROW ? __cursor_row_search(cbt, false, NULL, NULL) :
-                                      __cursor_col_search(cbt, NULL, NULL));
+    WT_RET(btree->type == BTREE_ROW ? __cursor_row_search(cbt, false, cbt->ref,  &leaf_found) :
+                                      __cursor_col_search(cbt, cbt->ref,  &leaf_found));
```
Comment (agorrod) [inline src/btree/bt_cursor.c:644]: I don't think this is the correct usage of row search with a starting position. Maybe this should be a call to {{__wt_btcur_search}} instead?

**Review by agorrod — COMMENTED** [review]
I see where this is going, and don't mind the idea of keeping a page pinned while it's relevant for prepared transaction resolution.
OTOH: I'm nervous about getting it right.
Could you craft a Python test that exercises the code path? Maybe using {{release_evict}} will help get the page evicted at the right time?

File: src/btree/bt_cursor.c (hunk 2 above)
Comment (luke-pearson) [reply src/btree/bt_cursor.c]: Done, see latest code.

**Comment by luke-pearson** [comment]
> Could you craft a Python test that exercises the code path?
I think we have numerous python tests that exercise that path, the difficulty here will be pausing during prepared resolution in a meaningful way. I think a timing stress could work potentially, I'll investigate.

File: src/btree/bt_cursor.c [hunk 3]
```diff
@@ -628,33 +628,37 @@ __wt_btcur_search_prepared(WT_CURSOR *cursor, WT_UPDATE **updp)
...
+     * Set the key only flag to indicate to the search that we don't want to check visibility we
+     * just want to position on a key. This short circuits validity checking.
      */
-    WT_RET(__cursor_reset(cbt));
-
-    WT_RET(btree->type == BTREE_ROW ? __cursor_row_search(cbt, false, NULL, NULL) :
-                                      __cursor_col_search(cbt, NULL, NULL));
-
+    F_SET(&cbt->iface, WT_CURSTD_KEY_ONLY);
+    ret = __wt_btcur_search(cbt);
```
Comment (quchenhao) [inline src/btree/bt_cursor.c:643]: Can you add a comment here to indicate we will search the pinned page first?

**Comment by quchenhao** [comment]
Can you run a round of stress test mdb tests? @luke-pearson

File: src/btree/bt_cursor.c (hunk 3 above)
Comment (luke-pearson) [reply src/btree/bt_cursor.c:647]: 6e05ab85e

**Review by quchenhao — APPROVED** [review]
lgtm

**Review by agorrod — APPROVED** [review]
I didn't think I was ready to approve this, but it's turned into a very nice change. Any time we get rid of conditional calls to row/column code I'm happy.
Thanks Luke

File: src/txn/txn.c [hunk 4]
```diff
@@ -1663,6 +1663,9 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
                  */
                 if (!F_ISSET(op, WT_TXN_OP_KEY_REPEATED))
                     WT_ERR(__txn_resolve_prepared_op(session, op, true, &cursor));
+
+                /* Sleep between resolving prepared operations when configured. */
+                __wt_timing_stress(session, WT_TIMING_STRESS_PREPARE_RESOLUTION, NULL);
```
Comment (kommiharibabu) [inline src/txn/txn.c:1668]: Just to confirm my understanding, this timing stress doesn't generate this failure as we do not release the pinned page at this location. But this change is useful to simulate longer prepare resolution.

**Review by kommiharibabu — APPROVED** [review]
LGTM

File: src/txn/txn.c (hunk 4 above)
Comment (luke-pearson) [reply src/txn/txn.c:1668]: It's not going to find this bug, as this bug is now fixed but it could find bugs like it in the future.

**Comment by luke-pearson** [comment]
The wiredtiger stress build and mongodb patch build look clean, merging.

**Comment by luke-pearson** [comment]
evergreen merge

**Pull request merged**